integration/upload_test_util.go

```diff
@@ -0,0 +1,200 @@
+		suite.T().Logf("file upload request: %s, with correlation id: %s", path, uploadRequest.CorrelationID)
+		requestedFiles[uploadRequest.CorrelationID] = path
+	}
+	suite.T().Logf("%v file upload requests, initiating uploads", len(requestedFiles))
+	return requestedFiles
+}
+
+// StartUploads starts the uploads for all given file upload requests
+func (suite *FileUploadSuite) StartUploads(featureID string, upload Upload, requestedFiles map[string]string) {
+	pathLastUpload := util.GetFeaturePropertyPath(featureID, propertyLastUpload)
+	topicCreated := util.GetTwinEventTopic(suite.ThingCfg.DeviceID, protocol.ActionCreated)
+	topicModified := util.GetTwinEventTopic(suite.ThingCfg.DeviceID, protocol.ActionModified)
+
+	connEvents, err := util.NewDigitalTwinWSConnection(suite.Cfg)
+	require.NoError(suite.T(), err, msgFailedCreateWebsocketConnection)
+	defer connEvents.Close()
+
+	err = util.SubscribeForWSMessages(suite.Cfg, connEvents, util.StartSendEvents, fmt.Sprintf(eventFilterTemplate, featureID))
+	require.NoError(suite.T(), err, "error subscribing for WS ditto events")
+	defer util.UnsubscribeFromWSMessages(suite.Cfg, connEvents, util.StopSendEvents)
+
+	for startID, path := range requestedFiles {
+		_, err := util.ExecuteOperation(suite.Cfg, suite.FeatureURL, operationStart, upload.requestUpload(startID, path))
+		require.NoErrorf(suite.T(), err, msgErrorExecutingOperation, operationStart)
+	}
+
+	statuses := []interface{}{}
+	err = util.ProcessWSMessages(suite.Cfg, connEvents,
+		func(msg *protocol.Envelope) (bool, error) {
+			if (msg.Topic.String() == topicCreated || msg.Topic.String() == topicModified) && msg.Path == pathLastUpload {
+				statuses = append(statuses, msg.Value)
+				return Contains(msg.Value, client.StateSuccess, client.StateFailed, client.StateCanceled), nil
+			}
+			return true, fmt.Errorf(msgUnexpectedValue, msg.Value)
+		})
+	require.NoError(suite.T(), err, "error processing upload status events")
+
+	lastUploadProgress := 0
+	for ind, status := range statuses {
+		uploadStatus := uploadStatus{}
+		err := util.Convert(status, &uploadStatus)
+		require.NoErrorf(suite.T(), err, "cannot convert %v to upload status", status)
+		suite.T().Logf("upload status event(%v)", uploadStatus)
+		require.GreaterOrEqual(suite.T(), uploadStatus.Progress, lastUploadProgress,
+			"upload status progress should be non-decreasing")
+		require.LessOrEqual(suite.T(), uploadStatus.Progress, 100,
+			"upload status progress should be less than or equal to 100%")
+		lastUploadProgress = uploadStatus.Progress
+		if ind < len(statuses)-1 {
+			require.Equal(suite.T(), client.StateUploading, uploadStatus.State, "wrong transitional upload state")
+		} else {
+			require.Equal(suite.T(), client.StateSuccess, uploadStatus.State, "wrong final upload state")
+		}
+	}
+}
+
+// Contains checks if a status "state" property is contained in the specified states
+func Contains(status interface{}, terminalStates ...string) bool {
```

Comment:
In my previous comment I suggested to use something more generic just because that's the implementation. Some more repairs are required:
- terminalStates - it should be: states
- terminal - it should be: state
- the method usage has to be double checked, is it enough to be `Contains` or it should be: `ContainsState`